test/e2e/network/dual_stack.go

```diff
@@ -0,0 +1,300 @@
+								Image:   imageutils.GetE2EImage(imageutils.Hostexec),
+								Command: []string{"sleep", "3600"},
+							},
+						},
+						Affinity: &v1.Affinity{
+							PodAntiAffinity: &v1.PodAntiAffinity{
+								RequiredDuringSchedulingIgnoredDuringExecution: []v1.PodAffinityTerm{
+									{
+										LabelSelector: &metav1.LabelSelector{
+											MatchExpressions: []metav1.LabelSelectorRequirement{
+												{
+													Key:      "test",
+													Operator: metav1.LabelSelectorOpIn,
+													Values:   []string{"dualstack-test-client"},
+												},
+											},
+										},
+										TopologyKey: "kubernetes.io/hostname",
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+
+		serverDeployment, err := cs.AppsV1().Deployments(f.Namespace.Name).Create(serverDeploymentSpec)
+		framework.ExpectNoError(err)
+
+		clientDeployment, err := cs.AppsV1().Deployments(f.Namespace.Name).Create(clientDeploymentSpec)
+		framework.ExpectNoError(err)
+
+		err = e2edeploy.WaitForDeploymentComplete(cs, serverDeployment)
+		framework.ExpectNoError(err)
+		err = e2edeploy.WaitForDeploymentComplete(cs, clientDeployment)
+		framework.ExpectNoError(err)
+
+		serverPods, err := e2edeploy.GetPodsForDeployment(cs, serverDeployment)
+		framework.ExpectNoError(err)
+
+		clientPods, err := e2edeploy.GetPodsForDeployment(cs, clientDeployment)
+		framework.ExpectNoError(err)
+
+		assertNetworkConnectivity(f, *serverPods, *clientPods)
+	})
+})
+
+var (
+	duration     = "10s"
+	pollInterval = "1s"
+	timeout      = 10
+)
+
+func assertNetworkConnectivity(f *framework.Framework, serverPods v1.PodList, clientPods v1.PodList) {
+	// curl from each client pod to all server pods to assert connectivity
+	var serverIPs []string
+	for _, pod := range serverPods.Items {
+		if pod.Status.PodIPs == nil || len(pod.Status.PodIPs) != 2 {
+			framework.Failf("PodIPs list not expected value, got %v", pod.Status.PodIPs)
+		}
+		if isIPv4(pod.Status.PodIPs[0].IP) == isIPv4(pod.Status.PodIPs[1].IP) {
+			framework.Failf("PodIPs should belong to different families, got %v", pod.Status.PodIPs)
+		}
+		serverIPs = append(serverIPs, pod.Status.PodIPs[0].IP, pod.Status.PodIPs[1].IP)
+	}
+
+	for _, clientPod := range clientPods.Items {
+		for _, ip := range serverIPs {
+			gomega.Consistently(func() error {
+				ginkgo.By(fmt.Sprintf("checking connectivity from pod %s to server ip %s", clientPod.Name, ip))
+				cmd := checkNetworkConnectivity(ip, "80")
+				_, _, err := f.ExecCommandInContainerWithFullOutput(clientPod.Name, "dualstack-test-client", cmd...)
+				return err
+			}, duration, pollInterval).ShouldNot(gomega.HaveOccurred())
+		}
+	}
+}
+
+func checkNetworkConnectivity(ip, port string) []string {
+	curl := fmt.Sprintf("curl -g --connect-timeout %v http://%s", 10, net.JoinHostPort(ip, port))
+	cmd := []string{"/bin/sh", "-c", curl}
+	return cmd
+}
+
+// isIPv4 checks if the provided ip belongs to ipv4 family.
```

Comment:
doesn't this function already exist in one of the util packages? ditto for the CIDR method below

Reply:
yup `k8s/utils` already vendor-ed on head.

Reply:
heh that change recently and only IPv6

can we at least TODO putting these in https://github.com/kubernetes/utils/blob/master/net/net.go instead?

Reply:
Done

Reply:
thanks!